src/crypto/index.js

```diff
@@ -363,41 +487,99 @@ Crypto.prototype.checkPrivateKey = function(privateKey, expectedPublicKey) {
+ * Args:
+ *     {function} A function that makes the request requiring auth. Receives the
+ *     auth data as an object.
+ */
+Crypto.prototype.resetCrossSigningKeys = async function(level, {
+    authUploadDeviceSigningKeys = async func => await func(),
+} = {}) {
+    logger.info(`Resetting cross-signing keys at level ${level}`);
+    // Copy old keys (usually empty) in case we need to revert
+    const oldKeys = Object.assign({}, this._crossSigningInfo.keys);
+    try {
+        await this._crossSigningInfo.resetKeys(level);
+        await this._signObject(this._crossSigningInfo.keys.master);
+
+        // send keys to server first before storing as trusted locally
+        // to ensure upload succeeds
+        const keys = {};
+        for (const [name, key] of Object.entries(this._crossSigningInfo.keys)) {
+            keys[name + "_key"] = key;
+        }
+        await authUploadDeviceSigningKeys(async authDict => {
+            await this._baseApis.uploadDeviceSigningKeys(authDict, keys);
+        });
+
+        // write a copy locally so we know these are trusted keys
+        await this._cryptoStore.doTxn(
+            'readwrite', [IndexedDBCryptoStore.STORE_ACCOUNT],
+            (txn) => {
+                this._cryptoStore.storeCrossSigningKeys(txn, this._crossSigningInfo.keys);
+            },
+        );
+    } catch (e) {
+        // If anything failed here, revert the keys so we know to try again from the start
+        // next time.
+        logger.error("Resetting cross-signing keys failed, revert to previous keys", e);
+        this._crossSigningInfo.keys = oldKeys;
+        throw e;
+    }
+    this._baseApis.emit("crossSigning.keysChanged", {});
+    await this._afterCrossSigningLocalKeyChange();
+    logger.info("Cross-signing key reset complete");
+};
 
-    // send keys to server
-    const keys = {};
-    for (const [name, key] of Object.entries(this._crossSigningInfo.keys)) {
-        keys[name + "_key"] = key;
+/**
+ * If cross-signing keys are known to exist in secret storage, this will get
+ * them and store them on this device as trusted.
+ */
+Crypto.prototype._getCrossSigningKeysFromSecretStorage = async function() {
+    logger.info("Getting cross-signing keys from secret storage");
+    // Copy old keys (usually empty) in case we need to revert
+    const oldKeys = Object.assign({}, this._crossSigningInfo.keys);
+    try {
+        await this._crossSigningInfo.getFromSecretStorage(this._secretStorage);
+        // XXX: Do we also need to sign the cross-signing master key with the
+        // device key as in `resetCrossSigningKeys`?
+
+        // write a copy locally so we know these are trusted keys
+        await this._cryptoStore.doTxn(
+            'readwrite', [IndexedDBCryptoStore.STORE_ACCOUNT],
+            (txn) => {
+                this._cryptoStore.storeCrossSigningKeys(txn, this._crossSigningInfo.keys);
+            },
+        );
```

Comment:
This one looks like it's also doing things that `checkOwnCrossSigningTrust()` ought to be doing (ie. the `this._cryptoStore.storeCrossSigningKeys` part).

Reply:
Right, I have converted to that.